folktables/utils/download_utils.py

```diff
@@ -0,0 +1,135 @@
+    ----------
+    url : str
+        URL from where the data will be downloaded.
+    download_path : pathlib.Path
+        Path to where the downloaded content will be stored.
+
+    Returns
+    -------
+    download_path : str
+        Path to where the content is stored.
+
+    Raises
+    ------
+    exceptions.FileDownloadError
+        This exception is raised if we get an error from our HTTP request
+        (i.e., the status code we get from the response is not 200).
+    """
+    response = requests.get(url)
+
+    if response.status_code != 200:
+        raise exceptions.FileDownloadError(
+            f'Failed to download the data from: {url}\n'
+            f'The HTTP request returned a {response.status_code} status code.'
+        )
+
+    with open(download_path, 'wb') as handle:
+        handle.write(response.content)
+
+    return download_path
+
+
+def async_download_file(files_resource):
+    """Wrapper around `download_file` function. This function allows us to
+    get the `url` and `download_path` from a `files_resource` and pass them
+    to the `download_file` function.
+    """
+    return download_file(files_resource.url,
+                         files_resource.download_path)
+
+
+def determine_files_to_download(files_resources, download, make_dir=True):
+    """Determines which datasets should be downloaded based on whether they
+    currently exist in local memory at the specified location.
+
+    Parameters
+    ----------
+    files_resources : list[FilesResource]
+        The FilesResource for the files the user has requested to (down)load.
+    download : bool
+        Whether the dataset should be downloaded or not.
+    make_dir : bool
+        Flag to determine whether we should attempt to create a directory
+        for every file that has been requested by the user.
+
+    Returns
+    -------
+    files_to_download : list[FilesResource]
+        The datasets that should be downloaded.
+
+    Raises
+    ------
+    FileNotFoundError
+        This error is raised when a file is not found in local memory and
+        the user has set the `download` flag to `False`.
+    """
+    files_to_download = []
+    for resource in files_resources:
+        if make_dir:
+            pathlib.Path(resource.data_dir).mkdir(exist_ok=True)
+
+        if resource.file_path.is_file():
+            continue
+
+        if not download:
+            raise FileNotFoundError(
+                f'Could not find the file '
+                f'{resource.file_name}. Call `get_data` '
+                f'with `download=True` to download the '
+                f'dataset and corresponding schema.'
+            )
+
+        files_to_download.append(resource)
+
+    return files_to_download
+
+
+def download_datasets(files_to_download):
+    """Downloads the datasets from their corresponding websites.
+    We're currently using `asyncio` to download multiple files
+    asynchronously. This, however, relies on `asyncio`'s `to_thread`
```

Comment:
I may be missing something: where are we using `to_thread`?

Reply:
Good catch! I forgot to update this docstring (we're not using `asyncio` anymore, rather we're only using `concurrent.futures`), so I'll go ahead and delete that part.